File: Programmers/[17684] 압축/Johoseong/17684.py
def solution(msg):
    answer = []
    arr = list(msg)
    N = len(arr)
    dic, index = dict(), 26
    
    for i in range(1,27): # 사전 만들기
        dic[chr(64 + i)] = i

    s, e = 0, 1
    while True:
        cur = msg[s:e]

        if e > N: # 종료조건
            answer.append(dic[prev])
            break

        if cur in dic:
            prev = cur
            e += 1
            continue
        else: #사전에 없음
            answer.append(dic[prev])
            index += 1
            dic[cur] = index
            s = e - 1

    return answer

print(solution("KAKAO"))

File: Programmers/[17684] 압축/woong-jae/17684.js
function solution(msg) {
    const answer = [];
    
    const initial = "ABCDEFGHIJKLMNOPQRSTUVWXYZ".split("").map((char, index) => [char, index + 1]);
    const dic = new Map(initial)
    let maxLength = 1;
    
    while(msg.length) {
        for(let length = maxLength; length > 0; length--) {
            const candidate = msg.slice(0, length);
            if(!dic.has(candidate)) continue;
            // w 찾음
            answer.push(dic.get(candidate));
            msg = msg.slice(length);
            if(msg.length) {
                const newWord = candidate + msg[0];
                dic.set(newWord, dic.size + 1);
                maxLength = Math.max(maxLength, newWord.length);
            }
            break;
        }
    }
    
    return answer;
}